### ISSUE TITLE
How to execute only one lint

### ISSUE THREAD

**Issue by sancho20021** [trailofbits/dylint#1418]
I couldn't find a discussion / chat / email to ask, so I am writing here :)

Is there a way to execute just ONE custom lint?
When I run `cargo dylint --path .` inside the lint library 18 lints are executed including some standard ones and one custom lint. Can I omit everything except my own lint?

Thanks in advance

**Comment by smoelius** [comment]
The answer depends on how the lint library is implemented.

It sounds like you have a library with more than one lint. In that case, the only way to run just one lint is to use [lint levels](https://doc.rust-lang.org/rustc/lints/levels.html), e.g., `DYLINT_RUSTFLAGS="-A lint1 -A lint2 ... -A lintn -W lint-of-interest" cargo dylint ...`. (You could use `RUSTFLAGS` instead of `DYLINT_RUSTFLAGS`, but it would cause all of the dependencies to be rebuilt.)

If having to write all of those flags sounds painful, it is.

We have the same need for the general-purpose and supplementary lints in Dylint. For this reason, the [`dylint_linting::declare_late_lint` ](https://docs.rs/dylint_linting/latest/dylint_linting/index.html#declare_late_lint-etc) macro provides a [`constituent` feature](https://docs.rs/dylint_linting/latest/dylint_linting/index.html#constituent-feature).

At a high level, the `constiuent` feature allows a lint to be built as an `rlib` to be linked into a larger library, or as a `cdylib` to be run by itself. For example, all of the general-purpose lints can be run with this command:
```sh
cargo dylint --git https://github.com/trailofbits/dylint --pattern examples/general
```
And just the `abs_home_path` lint can be run with this command:
```sh
cargo dylint --git https://github.com/trailofbits/dylint --pattern examples/general/abs_home_path
```

Admittedly, the `constiuent` feature is not as easy to use as I would like. The lint must have a feature to select building it as an `rlib` or a `cdylib` ([here in `abs_home_path`](https://github.com/trailofbits/dylint/blob/f4f281e593f7531820d66667d20ea7675a95cc4c/examples/general/abs_home_path/Cargo.toml#L24), for example).

Please let me know if I have not answered or misinterpreted your question.

**Comment by sancho20021** [comment]
Hello @smoelius . Thank you for responding! I think my library (I slightly modified the template which is created using `cargo dylint new ...`) has only one lint.
Here is the source code:
```rust
#![feature(rustc_private)]
#![feature(let_chains)]
#![warn(unused_extern_crates)]

extern crate rustc_hir;
extern crate rustc_span;

use clippy_utils::diagnostics::span_lint;
use rustc_lint::{LateContext, LateLintPass};
use rustc_span::Symbol;

dylint_linting::declare_late_lint! {
       pub KEK_VARIABLE_LINT,
    Warn,
    "description goes here"
}

impl<'tcx> LateLintPass<'tcx> for KekVariableLint {
    fn check_pat(&mut self, cx: &LateContext<'tcx>, pattern: &'tcx rustc_hir::Pat<'tcx>) {
        if let rustc_hir::PatKind::Binding(_, _, ident, _) = pattern.kind {
            if ident.name == Symbol::intern("kek") {
                println!("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!");
                span_lint(
                    cx,
                    KEK_VARIABLE_LINT,
                    ident.span,
                    "variable named `kek` is discouraged",
                );
            }
        }
    }
}

fn ui() {
    let kek = 3;
}
```

But when I run `cargo dylint --path .` inside the lint library, I get three warnings, not just one:
```
Checking with toolchain `nightly-2024-09-05-x86_64-unknown-linux-gnu`
warning: unused variable: `kek`
  --> src/lib.rs:35:9
   |
35 |     let kek = 3;
   |         ^^^ help: if this is intentional, prefix it with an underscore: `_kek`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: function `ui` is never used
  --> src/lib.rs:34:4
   |
34 | fn ui() {
   |    ^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: variable named `kek` is discouraged
  --> src/lib.rs:35:9
   |
35 |     let kek = 3;
   |         ^^^
   |
   = note: `#[warn(kek_variable_lint)]` on by default

warning: `new_lint_name` (lib) generated 3 warnings
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.10s
```

It includes my lint (about discouraged name for a variable) and two standard lints.
Is there a way to launch just my lint?

**Comment by smoelius** [comment]
Ah. Please try this:
```sh
RUSTFLAGS='--allow unused_variables' cargo dylint --path .
```

**Comment by sancho20021** [comment]
Okay, I see this logic. Would have to allow all other possible lints though. But thank you.
Anyway, my use case (not from the example I provided) is very strange thereby there is no necessity of supporting that. You can close the issue :)

**Comment by smoelius** [comment]
> Would have to allow all other possible lints though.

Right, which was why I said this can be painful. Just to be clear, there are other ways of allowing lints:
- in source code, e.g., but putting `#![allow(lint_name)]` at the top of a file
- in a Cargo.toml file by having a [lints section](https://doc.rust-lang.org/cargo/reference/manifest.html#the-lints-section), e.g.:
  ```
  [lints.clippy]
  pedantic = { level = "warn", priority = -1 }
  format-collect = "allow"
  missing-errors-doc = "allow"
  missing-panics-doc = "allow"
  ```
  (Copied from [here](https://github.com/trailofbits/cargo-unmaintained/blob/master/Cargo.toml#L62-L66). Sorry if you know all this already.)

I'll close the issue as you suggested. But please let me know if I can be of any more help.
